### PR TITLE
use vectors of pointers in middle-ram to reduce memory footprint

### LINKED ISSUE #381: Non-slim mode uses too much memory
8d803dc4d72f155dd89968916da12de297280c7f introduced a regression for middle-ram where it needs much more memory for its management structs than before. I'm no longer able to import Switzerland on a 8GB machine in non-slim mode.

I can provide a fix that reverts to having pointer arrays with dynamically allocated vectors (at the price of being marginally slower) but this will require C++11, so the libosmium branch needs to come first.

### PR DESCRIPTION
**Pull request by lonvia**
fixes #381

### CODE REVIEW

File: middle-ram.hpp [hunk 1]
```diff
@@ -61,15 +61,19 @@ struct middle_ram_t : public middle_t {
...
     struct ramRel {
         taglist_t tags;
         memberlist_t members;
+
+        ramRel(const taglist_t &t, const memberlist_t &m) : tags(t), members(m) {}
     };
 
-    std::vector<std::vector<ramWay> > ways;
-    std::vector<std::vector<ramRel> > rels;
+    std::vector<std::unique_ptr<std::vector<std::unique_ptr<ramWay> > > > ways;
```
Comment (pnorman) [inline middle-ram.hpp]: this is a vector of pointers to vectors of pointers to ways?

I'm wondering if it makes sense to typedef `std::vector<std::unique_ptr<ramWay> >` for clarity that it's a block of ways

**Comment by pnorman** [comment]
Do you think this needs any full planet testing? It looks okay to me.

**Comment by lonvia** [comment]
I doubt it's possible to still run non-slim on a full planet.

Redoing this with proper classes right now.

**Comment by pnorman** [comment]
> I doubt it's possible to still run non-slim on a full planet

Oh yes, good point.

**Comment by lonvia** [comment]
This is ready now. Tested with `./osm2pgsql -O null -C 16000 germany.osm.pbf`.

**Pull request merged**